### PR TITLE
feat(http): Support for per-host http concurrency limits

### LINKED ISSUE #6358: Support limiting concurrent requests per host
**What would you like Renovate to be able to do?**

Provide a configurable way and sensible defaults for concurrent limits per host.

**Describe the solution you'd like**

Implement it in the util/http logic. Mostly it should just work, but users can tweak it using hostRules. 

**Describe alternatives you've considered**

Today we try to rate limit by manager and package file, etc. It would be easier to have full concurrency of those and rely on the http layer to do rate limiting instead.

### PR DESCRIPTION
**Pull request by zharinov**
Closes #6358

### CODE REVIEW

File: lib/util/http/host-rules.ts [hunk 1]
```diff
@@ -37,3 +37,12 @@ export function applyHostRules(url: string, inOptions: GotOptions): GotOptions {
   }
   return options;
 }
+
+export function getRequestLimit(url: string): number | null {
+  const hostRule =
+    hostRules.find({
+      url,
+    }) || /* istanbul ignore next: can only happen in tests */ {};
```
Comment (viceice) [inline lib/util/http/host-rules.ts:45]: We should fix tests instead of using the fallback

File: lib/util/http/index.spec.ts [hunk 2]
```diff
@@ -131,4 +134,73 @@ describe(getName(__filename), () => {
...
+      return [request, resolveRequest, response, resolveResponse];
+    };
+
+    const [fooReq, fooStart, fooResp, fooFinish] = mockRequestResponse();
+    const [barReq, barStart, barResp, barFinish] = mockRequestResponse();
+
+    nock(baseUrl)
+      .get('/foo')
+      .reply(200, async () => {
+        await delay(100);
```
Comment (viceice) [inline lib/util/http/index.spec.ts:165]: using nock delay instead?

 https://github.com/nock/nock#delay-the-response

File: lib/workers/repository/stats.ts [hunk 3]
```diff
@@ -17,26 +17,39 @@ export function printRequestStats(): void {
     return 1;
   });
   const allRequests: string[] = [];
-  const requestHosts: Record<string, number[]> = {};
-  for (const request of httpRequests) {
+  const requestHosts: Record<
+    string,
+    { duration: number; queueDuration: number }[]
```
Comment (viceice) [inline lib/workers/repository/stats.ts:22]: We should extract module local type here

File: lib/util/http/queue.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    return URL.parse(url).host;
+  } catch (e) /* istanbul ignore next */ {
+    return null;
+  }
+}
+
+export function getQueue(url: string): PQueue | null {
+  const host = getUrlHost(url);
+
+  /* istanbul ignore if */
```
Comment (viceice) [inline lib/util/http/queue.ts:18]: It's simple, but should be tested trough

File: lib/util/http/queue.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import URL from 'url';
+import PQueue from 'p-queue';
+import { getRequestLimit } from './host-rules';
+
+let hostQueues: Record<string, PQueue> = {};
```
Comment (viceice) [inline lib/util/http/queue.ts:5]:
```suggestion
let hostQueues: Record<string, PQueue> = Object.create({});
```
So we don't inherit object prototype.

We alternatively can use a `Map<string, PQueue>`, so we can make it `const` and call `hostQueues.clear()` to cleanup

File: lib/util/http/queue.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  let queue = hostQueues[host];
+  if (!queue) {
+    const concurrency = getRequestLimit(url);
+    if (concurrency) {
+      queue = new PQueue({ concurrency });
+      hostQueues[host] = queue;
+    } else {
+      queue = null;
+    }
+  }
```
Comment (viceice) [inline lib/util/http/queue.ts:32]:
```suggestion
  if (queue === undefined) {
    queue = null;
    const concurrency = getRequestLimit(url);
    if (concurrency) {
      queue = new PQueue({ concurrency });
    }
  }
  hostQueues[host] = queue;
```
Optimize, so we only check once for limit

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/http/queue.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+export function getQueue(url: string): PQueue | null {
+  const host = getUrlHost(url);
+  if (!host) {
+    return null;
+  }
+
+  let queue = hostQueues.get(host);
+  if (queue === undefined) {
+    queue = null;
```
Comment (rarkins) [inline lib/util/http/queue.ts:23]:
```suggestion
    queue = null; // null represents "no queue", as opposed to undefined
```

File: lib/util/http/queue.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    if (concurrency) {
+      queue = new PQueue({ concurrency });
+    }
+  }
+  hostQueues.set(host, queue);
+
+  return queue;
+}
+
+export function clear(): void {
```
Comment (rarkins) [inline lib/util/http/queue.ts:34]: This should be called before every repository, right?

File: lib/util/http/index.ts [hunk 9]
```diff
@@ -126,20 +127,34 @@ export class Http<GetOptions = HttpOptions, PostOptions = HttpPostOptions> {
...
+          duration,
+          queueDuration,
+        });
+        memCache.set('http-requests', httpRequests);
+        return resp;
+      });
+    };
+
+    const queue = getQueue(url);
+    startTime = Date.now();
```
Comment (rarkins) [inline lib/util/http/index.ts:150]: Is there the possibility that the startTime value could be overwritten by concurrent requests?

File: lib/util/http/index.ts [hunk 10]
```diff
@@ -126,20 +127,34 @@ export class Http<GetOptions = HttpOptions, PostOptions = HttpPostOptions> {
         return resolveResponse<T>(cachedRes, options);
       }
     }
-    const startTime = Date.now();
-    const promisedRes = got<T>(url, options);
+    let startTime: number;
+    const queueTask = (): Promise<Response<T>> => {
+      const queueTime = Date.now();
+      const queueDuration = queueTime - startTime;
+      return got<T>(url, options).then((resp) => {
```
Comment (rarkins) [inline lib/util/http/index.ts:134]: We have very little remaining `.then()` use in the repo. not possible to structure this as async/await?

File: lib/util/http/index.ts (hunk 9 above)
Comment (zharinov) [reply lib/util/http/index.ts:150]: I don't think so, because we create `startTime` closure over corresponding `queueTask()` in each request.

File: lib/util/http/queue.ts (hunk 8 above)
Comment (zharinov) [reply lib/util/http/queue.ts:34]: Added before `try-catch` block in `renovateRepo()`

File: lib/util/http/index.ts (hunk 9 above)
Comment (zharinov) [reply lib/util/http/index.ts]: Though, I made this code simpler

File: lib/util/http/index.ts [hunk 11]
```diff
@@ -80,6 +81,25 @@ function applyDefaultHeaders(options: Options): void {
   };
 }
 
+async function gotTask<T>(
```
Comment (rarkins) [inline lib/util/http/index.ts:84]: Can this refactoring be done in a separate PR now?

Comment (zharinov) [reply lib/util/http/index.ts:84]: We don't have `.then(...)` calls in `master` branch. Or do you mean something else?

Comment (rarkins) [reply lib/util/http/index.ts:84]: I was wondering how much of the refactoring we can do separately before adding the queue. Or do you think it's not really possible?

Comment (rarkins) [reply lib/util/http/index.ts:84]: I thought this function looked like it could be refactored independently of the queuing

Comment (zharinov) [reply lib/util/http/index.ts:84]: Let's see what can I extract...

**Comment by zharinov** [comment]
Just received stats like this with queue of size 3:
```
DEBUG: http statistics (repository=zharinov/renovate-test-got-queue)
       "hostStats": [
         "api.github.com, 4 requests, 387ms request average, 0ms queue average",
         "repo.maven.apache.org, 1259 requests, 383ms request average, 2540ms queue average"
       ],
       "totalRequests": 1263
```

File: docs/usage/configuration-options.md [hunk 12]
```diff
@@ -633,6 +633,10 @@ Renovate does not do a "longest match" algorithm to pick between multiple matchi
 
 If you have any uncertainty about exactly which hosts a service uses, then it can be more reliable to use `domainName` instead of `hostName` or `baseUrl`. e.g. configure `"hostName": "docker.io"` to cover both `index.docker.io` and `auth.docker.io` and any other host that's in use.
 
+### concurrentRequestLimit
+
+Limit number of concurrent requests on per-host basis.
```
Comment (zharinov) [inline docs/usage/configuration-options.md:638]: Fixme

File: lib/util/http/queue.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+function getUrlHost(url: string): string | null {
+  try {
+    return URL.parse(url).host;
+  } catch (e) {
+    return null;
+  }
+}
+
+export function getQueue(url: string): PQueue | null {
+  const host = getUrlHost(url);
```
Comment (zharinov) [inline lib/util/http/queue.ts:16]: If keys are just hosts, here is possible problem if we're processing multiple repos per run, isn't it?

Comment (zharinov) [reply lib/util/http/queue.ts:16]: Never mind, we clear queues

File: docs/usage/configuration-options.md [hunk 14]
```diff
@@ -729,6 +729,10 @@ Renovate does not do a "longest match" algorithm to pick between multiple matchi
 If you have any uncertainty about exactly which hosts a service uses, then it can be more reliable to use `domainName` instead of `hostName` or `baseUrl`.
 e.g. configure `"hostName": "docker.io"` to cover both `index.docker.io` and `auth.docker.io` and any other host that's in use.
 
+### concurrentRequestLimit
+
+Limit number of concurrent requests on per-host basis.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:734]:
```suggestion
Setting this value will limit the number of concurrent outstanding requests for any host it applies to. It should be used only if a datasource is rate limiting Renovate or having problems with the load.
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md [hunk 15]
```diff
@@ -729,6 +729,10 @@ Renovate does not do a "longest match" algorithm to pick between multiple matchi
 If you have any uncertainty about exactly which hosts a service uses, then it can be more reliable to use `domainName` instead of `hostName` or `baseUrl`.
 e.g. configure `"hostName": "docker.io"` to cover both `index.docker.io` and `auth.docker.io` and any other host that's in use.
 
+### concurrentRequestLimit
+
+Setting this value will limit the number of concurrent outstanding requests for any host it applies to. It should be used only if a datasource is rate limiting Renovate or having problems with the load.
```
Comment (viceice) [inline docs/usage/configuration-options.md:734]: @HonkingGoose Can you check this please.

File: lib/util/http/index.ts [hunk 16]
```diff
@@ -69,20 +70,23 @@ function applyDefaultHeaders(options: Options): void {
...
   const resp = await got<T>(url, options);
   const responseTime = Date.now();
+
   const httpRequests = memCache.get('http-requests') || [];
   httpRequests.push({
     method: options.method,
     url,
     duration: responseTime - requestTime,
-    queueDuration: requestTime - startTime,
+    queueDuration,
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/util/http/index.ts:86]: Isn't this queuing done outside/prior to got?

File: docs/usage/configuration-options.md (hunk 15 above)
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:734]:
```suggestion
Usually the default `concurrentRequestLimit` is fine, but you can use `concurrentRequestLimit` to limit the number of concurrent outstanding requests.
You only need to adjust this setting if a datasource is rate limiting Renovate or has problems with the load.
The limit will be set for any host it applies to.
```

How about something like this?

Comment (zharinov) [reply docs/usage/configuration-options.md:734]: @HonkingGoose I like this forumlation, also I'll add example snippet

File: docs/usage/configuration-options.md [hunk 17]
```diff
@@ -724,6 +724,26 @@ For example, `"baseUrl": "https://api.github.com"` is equivalent to `"hostName":
 
 Renovate does not do a "longest match" algorithm to pick between multiple matching `baseUrl` values in different rules, so put the longer `baseUrl` rule _after_ the shorter one in your `hostRules`.
 
+### concurrentRequestLimit
+
+Usually the default setting is fine, but you can use `concurrentRequestLimit` to limit the number of concurrent
+outstanding requests.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:730]:
```suggestion
Usually the default setting is fine, but you can use `concurrentRequestLimit` to limit the number of concurrent outstanding requests.
```

Fix so that we conform to the "one sentence per line" system.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 17 above)
Comment (zharinov) [reply docs/usage/configuration-options.md:730]: Ouch, that's my code reformatting

File: docs/usage/configuration-options.md [hunk 18]
```diff
@@ -724,6 +724,25 @@ For example, `"baseUrl": "https://api.github.com"` is equivalent to `"hostName":
 
 Renovate does not do a "longest match" algorithm to pick between multiple matching `baseUrl` values in different rules, so put the longer `baseUrl` rule _after_ the shorter one in your `hostRules`.
 
+### concurrentRequestLimit
+
+Usually the default setting is fine, but you can use `concurrentRequestLimit` to limit the number of concurrent outstanding requests.
+You only need to adjust this setting if a datasource is rate limiting Renovate or has problems with the load.
+The limit will be set for any host it applies to.
+
+Example:
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:733]:
```suggestion
e.g.:
```

@rarkins do you like `Example:` or `e.g.:` better here?

Comment (rarkins) [reply docs/usage/configuration-options.md:733]: I think maybe "Example config:" is more verbose and readable

**Comment by rarkins** [comment]
@zharinov I think a good test would be a repo with >20 dependencies and then limit registry.npmjs.org to 2 concurrent requests

File: docs/usage/configuration-options.md (hunk 18 above)
Comment (HonkingGoose) [reply docs/usage/configuration-options.md:733]:
```suggestion
Example config:
```

Let's go with this then. 😄

**Review by HonkingGoose — APPROVED** [review]

**Comment by viceice** [comment]
> @zharinov I think a good test would be a repo with >20 dependencies and then limit registry.npmjs.org to 2 concurrent requests

Maybe ours (forked of cause) 😏

**Comment by zharinov** [comment]
Here is the reason I'm not receiving PR from `16.04` to `20.04`:
https://github.com/renovatebot/renovate/blob/7cec2f6f703076918b907d84a46d6922273568fc/lib/datasource/docker/index.ts#L26

**Comment by rarkins** [comment]
Yes, fixing Ubuntu and Node versioning are our pre-requisites for removing that restriction :)

**Comment by zharinov** [comment]
Unlimited: `repo.maven.apache.org, 1281 requests, 181ms request average, 0ms queue average`
Limit = 2: `repo.maven.apache.org, 1281 requests, 169ms request average, 1650ms queue average`
Limit = 1: `repo.maven.apache.org, 1281 requests, 194ms request average, 4040ms queue average`

**Review by viceice — APPROVED** [review]
